src/middlewares/withAuth.ts

```diff
@@ -0,0 +1,188 @@
+  ) {
+    throw new Error('UAA environment variables are not set');
+  }
+
+  const logoutUrl = new URL(
+    process.env.UAA_ROOT_URL + process.env.UAA_LOGOUT_PATH
+  );
+  const params = new URLSearchParams(logoutUrl.search);
+  params.set('client_id', process.env.OAUTH_CLIENT_ID);
+  params.set('redirect', process.env.ROOT_URL + process.env.AUTH_CALLBACK_PATH);
+  const response = NextResponse.redirect(logoutUrl + '?' + params.toString());
+  response.cookies.delete('authsession');
+  return response;
+}
+
+export function setAuthCookie(
+  data: UAATokenResponseObj,
+  response: NextResponse
+) {
+  const decodedToken = decodeJwt(data.access_token);
+  response.cookies.set(
+    'authsession',
+    JSON.stringify({
+      accessToken: data.access_token,
+      user_id: decodedToken.user_id,
+      user_name: decodedToken.user_name,
+      email: decodedToken.email,
+      refreshToken: data.refresh_token,
+      expiry: Date.now() + data.expires_in * 1000,
+    })
+  );
+  return response;
+}
+
+export async function requestAndSetAuthToken(request: NextRequest) {
+  if (
+    !process.env.UAA_ROOT_URL ||
+    !process.env.OAUTH_CLIENT_ID ||
+    !process.env.OAUTH_CLIENT_SECRET
+  ) {
+    throw new Error('UAA environment variables are not set');
+  }
+
+  const stateCookie = request.cookies.get('state');
+  let response;
+  let lastPagePath;
+  if ((lastPagePath = request.cookies.get('last_page')?.value)) {
+    response = NextResponse.redirect(new URL(lastPagePath, request.url));
+  } else {
+    response = NextResponse.redirect(new URL('/', request.url));
+  }
+
+  if (
+    !stateCookie ||
+    request.nextUrl.searchParams.get('state') != stateCookie['value']
+  ) {
+    return response;
+  }
+  const data = await postToAuthTokenUrl({
+    code: request.nextUrl.searchParams.get('code') || '',
+    grant_type: 'authorization_code',
+    response_type: 'token',
+    client_id: process.env.OAUTH_CLIENT_ID,
+    client_secret: process.env.OAUTH_CLIENT_SECRET,
+  });
+  response = setAuthCookie(data, response);
+  response.cookies.delete('state');
+  response.cookies.delete('last_page');
+  return response;
+}
+
+export async function refreshAuthToken(refreshToken: string) {
+  if (!process.env.OAUTH_CLIENT_ID || !process.env.OAUTH_CLIENT_SECRET) {
+    throw new Error('OAUTH environment variables are not set');
+  }
+
+  const data = await postToAuthTokenUrl({
+    grant_type: 'refresh_token',
+    refresh_token: refreshToken,
+    client_id: process.env.OAUTH_CLIENT_ID,
+    client_secret: process.env.OAUTH_CLIENT_SECRET,
+  });
+  return data;
+}
+
+export function redirectToLogin(request: NextRequest): NextResponse {
+  const loginPath = logInPath();
+  const response = NextResponse.redirect(new URL(loginPath, request.url));
+  response.cookies.set('last_page', request.nextUrl.pathname);
+  return response;
+}
+
+export async function authenticateRoute(request: NextRequest) {
+  let response = NextResponse.next();
```

Comment:
[blocking question] My understanding of the stacked approach is that the same `response` object gets passed through and is modified by each middleware function. Does how we produce a response in each function matter? I see that this function uses `response = NextResponse.next()` whereas withCSP uses `response = await next(request, _next)`. Is one preferable over the other?

Reply:
We'll test this when deployed